File: modules/wem-api/src/main/java/com/enonic/wem/api/content/page/PageDescriptor.java
package com.enonic.wem.api.content.page;


import com.google.common.base.Preconditions;

import com.enonic.wem.api.content.page.region.RegionDescriptors;
import com.enonic.wem.api.form.Form;
import com.enonic.wem.api.module.ModuleResourceKey;

public final class PageDescriptor
{
    private final PageDescriptorKey key;

    private final String displayName;

    private final Form config;

    private final RegionDescriptors regions;

    private PageDescriptor( final Builder builder )
    {
        Preconditions.checkNotNull( builder.key, "key cannot be null" );
        Preconditions.checkNotNull( builder.config, "config cannot be null" );
        Preconditions.checkNotNull( builder.regions, "regions cannot be null" );
        this.key = builder.key;
        this.displayName = builder.displayName;
        this.regions = builder.regions;
        this.config = builder.config != null ? builder.config : Form.newForm().build();
    }

    public PageDescriptorKey getKey()
    {
        return key;
    }

    public ComponentDescriptorName getName()
    {
        return this.key.getName();
    }

    public ModuleResourceKey getResourceKey()
    {
        return ModuleResourceKey.from( key.getModuleKey(), "page/" + key.getName().toString() );
    }

    public String getDisplayName()
    {
        return displayName;
    }

    public Form getConfig()
    {
        return config;
    }

    public RegionDescriptors getRegions()
    {
        return regions;
    }

    public static PageDescriptor.Builder newPageDescriptor()
    {
        return new Builder();
    }

    public static class Builder
    {
        private PageDescriptorKey key;

        private String displayName;

        private Form config;

        private RegionDescriptors regions;

        private Builder()
        {
        }

        public Builder key( final PageDescriptorKey value )
        {
            this.key = value;
            return this;
        }

        public Builder displayName( final String value )
        {
            this.displayName = value;
            return this;
        }

        public Builder config( final Form value )
        {
            this.config = value;
            return this;
        }

        public Builder regions( final RegionDescriptors value )
        {
            this.regions = value;
            return this;
        }

        public PageDescriptor build()
        {
            return new PageDescriptor( this );
        }
    }
}

File: modules/wem-webapp/src/main/webapp/admin/common/js/content/page/PageDescriptor.ts
module api.content.page {

    export class PageDescriptor extends Descriptor {

        private regions: region.RegionDescriptor[];

        constructor(builder: PageDescriptorBuilder) {
            super(builder);
            this.regions = builder.regions;
        }

        public getRegions(): region.RegionDescriptor[] {
            return this.regions;
        }
    }

    export class PageDescriptorBuilder extends DescriptorBuilder {

        regions: region.RegionDescriptor[] = [];

        public fromJson(json: api.content.page.json.PageDescriptorJson): PageDescriptorBuilder {

            this.setName(new DescriptorName(json.name));
            this.setDisplayName(json.displayName);
            this.setConfig(json.config != null ? new api.form.Form(json.config) : null);
            for (var i = 0; i < json.regions.length; i++) {
                var region = new api.content.page.region.RegionDescriptorBuilder().fromJson(json.regions[i]).build();
                this.regions.push(region);
            }

            return this;
        }

        public setName(value: DescriptorName): PageDescriptorBuilder {
            this.name = value;
            return this;
        }

        public setDisplayName(value: string): PageDescriptorBuilder {
            this.displayName = value;
            return this;
        }

        public setConfig(value: api.form.Form): PageDescriptorBuilder {
            this.config = value;
            return this;
        }

        public addRegion(value: region.RegionDescriptor): PageDescriptorBuilder {
            this.regions.push(value);
            return this;
        }

        public build(): PageDescriptor {
            return new PageDescriptor(this);
        }
    }
}